Longstep with GMREDI

hajsong opened the pull request:
![x-unix-mode=0644](https://github.com/user-attachments/assets/2853f09e-f955-42ed-832b-4baadede6299)
## What changes does this PR introduce?
These changes include the GMREDI related terms in the LONGSTEP package.


## What is the current behaviour? 
Some of the GMREDI terms are not included in the LONGSTEP package when computing the transport


## What is the new behaviour 
Those terms are now included, and they are used in the GMREDI package when computing the transport of tracers.


## Does this PR introduce a breaking change? 
This PR improves the LONGSTEP performance when using GMREDI, as shown in the attached figure (RMSE of the passive tracer with and without this modification with respect to the online simulation).


## Other information:
I tested this changes in `global_ocean.cs32x15` but after including a passive tracer just as in `tutorial_global_oce_latlon`.

## Suggested addition to `tag-index`
(To avoid unnecessary merge conflicts, please don't update `tag-index`. One of the admins will do that when merging your pull request.)

jm-c commented:
@hajsong Thanks for sharing this. A side comment: I wanted to remove this "genmake_local" (from `global_ocean,cs32x15`) so it's good that you got rid off.

hajsong commented:
I forgot to mention that Jean-Michel pointed out this issue and guided me in making these changes. Thank you, Jean-Michel!

jahn commented:
This looks good.  Right now it does not compile with GM_BOLUS_ADVEC undefined.  I think the PsiX/Y bits need to be in #ifdef blocks, and maybe also `IF (GM_AdvForm)`.  I wonder whether we should always have the diagnostics defined, though, regardless of both GM_BOLUS_ADVEC and GM_EXTRA_DIAGONAL, and just not fill them.  This is what is done for the tensor components in the GMREDI package itself.

hajsong commented:
@jahn, thank you for catching that issue (and I am so sorry for the late response)! I will take a look at it.

hajsong commented:
I revisited this pull request and verified the issue pointed out by @jahn. So I added `#ifdef GM_BOLUS_ADVEC` in `longstep_average.F` to prevent the compiling error.
I also found out that the error occurred in `gmredi` package when compiling with `GM_BOLUS_ADVEC` undefined and `GM_BOLUS_BVP` defined. To resolve this, I modified `GMREDI.h` to force to turn on `GM_BOLUS_ADVEC` when `GM_BOLUS_BVP` is defined.

jahn commented on pkg/gmredi/GMREDI_OPTIONS.h:
```diff
@@ -56,10 +56,13 @@ C be used with the advective form (Bolus velocity) of GM
 C Allows to use the advective form (Bolus velocity) of GM
 C  instead of the Skew-Flux form (=default)
-#define GM_BOLUS_ADVEC
+#undef GM_BOLUS_ADVEC
```
Would you maybe prefer to keep GM_BOLUS_ADVEC defined by default (regardless of GM_BOLUS_BVP), @jm-c?

jahn reviewed (approved):
This looks good to me now.

jm-c commented:
@hajsong Thanks for this contribution. I took a look at changes, and have 2 suggestions:

1. Could simplify these pieces of code by removing the GM specific options that are set from `GMREDI_OPTIONS.h`, mainly `GM_EXTRA_DIAGONAL` from `LONGSTEP.h`. This way we don't need to include `GMREDI_OPTIONS.h` in each `*.F` file that contains `LONGSTEP.h`. This would also offer a symmetrical treatment of `GM_EXTRA_DIAGONAL` and `GM_BOLUS_ADVEC`. And finally, this would fix a problem in `longstep_average.F`, line 204, when `EXCH_UV_3D_RL` is applied to `LS_Kuz, LS_Kvz` even when `GM_EXTRA_DIAGONAL` is undef.
2. And regarding the changes to `GMREDI_OPTIONS.h` related to `GM_BOLUS_BVP`, I would prefer to keep `GMREDI_OPTIONS.h` unchanged and to apply a simple fix to `gmredi_calc_psi_bvp.F` so that it compiles. One thing that is missing (not from this PR) is few check and stop in `gmredi_check.F` for a few combinations of options and run-time switches, including those related to `GM_useBVP`.

I could take care of these two if you like, just let me know.

jm-c commented:
I found a bug, as `LS_PsiXCount` and `LS_PsiYCount` were declared in `LONGSTEP.h` but not stored in common block. WIll fix this as part of the other commits.

jm-c commented:
@hajsong I am done with my changes and just running few short `testreport`.

jm-c reviewed (approved):
LGTM my few `testreport` run turn good.

hajsong commented:
Thank you for going through the test, @jm-c !

jm-c commented:
@jahn Should I merge ?

jahn commented:
Yes, please!
